File: 021. Merge Two Sorted Lists/list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        if(l1 == nullptr) return l2;
        if(l2 == nullptr) return l1;
        ListNode* dummy = new ListNode(-1), *tail = dummy;
        while(l1 && l2){
            int a = l1->val, b = l2->val;
            if(a > b){
                tail->next = l2;
                tail = tail->next;
                l2 = l2->next;
            }else{
                tail->next = l1;
                tail = tail->next;
                l1 = l1->next;
            }
        }

        ListNode* rest = nullptr;
        if(l1) rest = l1;
        if(l2) rest = l2;
        while(rest){
            tail->next = rest;
            tail = tail->next;
            rest = rest->next;
        }

        return dummy->next;
    }
};

File: 0002.AddTwoNumbers/list.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode AddTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(-1), p = dummy;
        int carry = 0;
        while(l1 != null && l2 != null){
            int cur = l1.val + l2.val + carry;
            carry = cur / 10;
            cur = cur % 10;
            ListNode tmp = new ListNode(cur);
            p.next = tmp;
            p = p.next;
            l1 = l1.next;
            l2 = l2.next;
        }
        ListNode rest = null;
        if(l1 != null)
            rest = l1;
        if(l2 != null)
            rest = l2;
        while(rest != null){
            int cur = rest.val + carry;
            carry = cur / 10;
            cur = cur % 10;
            ListNode tmp = new ListNode(cur);
            p.next = tmp;
            p = p.next;
            rest = rest.next;
        }
        if(carry != 0){
            p.next = new ListNode(carry);
        }
        
        return dummy.next;
    }
}